### DISCUSSION TITLE
need help understanding aggregating index concept from SQL perspective

### DISCUSSION THREAD

**Discussion by alex-bermishev** [ravendb/ravendb#14740] [Q&A] [answered]
goal:
retrieve first and last timestamps of transactions for a given period (variable) of time grouped by transaction type

in plain SQL I can do this:

```SQL 
SELECT
	[TRANTYPE_ID],
	MIN( [TRAN_DATE] ),
	MAX( [TRAN_DATE] ),
	COUNT( * )
FROM [TRANSLOG]
WHERE ( [TRAN_DATE] BETWEEN '2015-01-01' AND '2020-01-01' )
GROUP BY
	[TRANTYPE_ID]
```

where date interval parameters are variable

I've learned that more of the same result can be achieved in RavenDB by creating indexes only

here are the questions:
- this query takes parameters from business layer, and RavenDB index cannot incorporate any variables. what is the best approach to implement map/reduce statements?
- I guess, I can create temporary index for every given timestamp range, but it will be truly `temporary`. what is the best approach to `garbage collecting` used once indexes?
- I believe I am really missing some core understanding of the whole concept =). maybe I am looking in the totally wrong direction. please, enlighten me with the best way to get this task done with RavenDB

thanks

----- added
it is possible to create `temporary single use` index with [PutIndexesOperation](https://ravendb.net/docs/article-page/5.0/csharp/client-api/operations/maintenance/indexes/put-indexes) and drop it with [DeleteIndexOperation](https://ravendb.net/docs/article-page/5.0/csharp/client-api/operations/maintenance/indexes/delete-index)

**Comment by ayende** [comment]
Using temporary indexes like that is **not** recommended
For flexible range like that, you can do that using facets.

Alternatively, if the data is timestamp base, we have direct support for this with time series data.

**Comment by alex-bermishev** [comment, marked as answer]
I've tried different approaches and explored possibilities of RavenDB
just in case it may be helpful for others
here is complete task
given large collection of events (5+ mln) of the following structure

```JSON
{
    "Header": {
        "ID": "1720fba6-4a99-4213-9414-9c23b886aa9a",
        "Timestamp": "2000-03-10T10:39:00.000+03:00",
        "User": "$root",
        "Sender": "test/territory/event_generator",
        "Name": "territory_exit",
        "Schema": "territory"
    },
    "Payload": {
        "territory_id": "02d57053-09b3-42d0-b205-b5d9b384bb88",
        "identifiers": [
            {
                "code": "",
                "subject_id": "f8f48061-5d5a-4070-847c-a839b6b425eb"
            },
            {
                "code": "",
                "subject_id": "1b90a88d-c486-4fec-a7fa-0ac03a5a6e3e"
            }
        ]
    }
}
```
task is to retrieve documents of the first entry and the last exit of each subject (identified by `Payload.identifiers[].subject_id`)
- for specified timestamp interval
- for specified subjects

@ayende noted that there is a possibility to utilize time series or facets query
none of this features really suit for this task
- time series is about scalar values, changing by time, attached to some document (does not fit, as we have stream of documents by time here)
- facets give possibility to aggregate scalar values, (does not fit, as we need documents in the result)

here is what I came up with for now
1. for each query create temporary index via `IndexDefinition` and send it to server like this
```C#
        var indexDefinition = new IndexDefinition
        {
            Name = $"events/{Guid.NewGuid()}",
            Maps =
            {
                @$"

                from e in docs.events
                from identifier in e.Payload.identifiers
                ....
                "
            },

            Reduce =
                @"

                from result in results
                group result by result.SubjectID into g
                ....
                
                "
        };

        documentStore.Maintenance
            .Send( new PutIndexesOperation( indexDefinition ) );
```
2. query temporary index like this
```C#
        List<JObject> result = default;
        using ( var session = documentStore.OpenSession() )
            result = session
                .Query<JObject>( indexDefinition.Name )
                .Customize( customization => customization.WaitForNonStaleResults() )                
                .ToList();
```

3. then delete temporary index like this
```C#
        documentStore.Maintenance
            .Send( new DeleteIndexOperation( indexDefinition.Name ) );
```

4. actual defined map-reduce index is the following
map:
```
from e in docs.events
from identifier in e.Payload.identifiers
let subjectIDs = new HashSet<string> 
{ 
    "57d68a46-8af1-4287-882e-e46ea75f6418", 
    "80a58d56-9081-45b4-9801-6f8be6951197" 
}
where 
    (e.Header.Timestamp >= DateTimeOffset.Parse("2000-01-01T00:00:00.000+03:00") 
    && e.Header.Timestamp <= DateTimeOffset.Parse("2000-01-05T00:00:00.000+03:00")) 
    && (e.Header.Name == "territory_entry" || e.Header.Name == "territory_exit") 
    && subjectIDs.Contains(identifier.subject_id)
select new
{
    SubjectID = identifier.subject_id,
    Entry = 
        e.Header.Name == "territory_entry" 
        ? new
        {
            ID = e.Header.ID,
            Name = e.Header.Name,
            Timestamp = e.Header.Timestamp,
            TerritoryID = e.Payload.territory_id,
            Identifier = identifier
        }
        : null,
    Exit = 
        e.Header.Name == "territory_exit" 
        ? new
        {
            ID = e.Header.ID,
            Name = e.Header.Name,
            Timestamp = e.Header.Timestamp,
            TerritoryID = e.Payload.territory_id,
            Identifier = identifier
        }
        : null
}
```

reduce:
```
from result in results
group result by result.SubjectID into g
select new
{
    SubjectID = g.Key,
    Entry = g
        .Where( x => x.Entry != null )
        .OrderBy( x => x.Entry.Timestamp )
        .FirstOrDefault()
        ?.Entry,
    Exit = g
        .Where( x => x.Exit.Name != null )
        .OrderByDescending( x => x.Exit.Timestamp )
        .FirstOrDefault()
        ?.Exit
}
```
5. result of this temporary index query is like this
```JSON
{
    "SubjectID": "57d68a46-8af1-4287-882e-e46ea75f6418",
    "Entry": {
        "ID": "a111aae2-b099-4fae-8b14-48ee675d60b7",
        "Name": "territory_entry",
        "Timestamp": "2000-01-01T04:08:00.0000000",
        "TerritoryID": "61fafd9d-6aa3-4da8-975d-e707a5aa0612",
        "Identifier": {
            "code": "",
            "subject_id": "57d68a46-8af1-4287-882e-e46ea75f6418"
        }
    },
    "Exit": {
        "ID": "f084f062-e435-4382-8ea0-d358cc5096d4",
        "Name": "territory_exit",
        "Timestamp": "2000-01-01T13:32:00.0000000",
        "TerritoryID": "cae6a730-6e9a-4046-af9d-0b17d258309d",
        "Identifier": {
            "code": "",
            "subject_id": "57d68a46-8af1-4287-882e-e46ea75f6418"
        }
    }
}
```

**Comment by alex-bermishev** [comment]
question:
is it possible to define index with TTL (time to live) which will be auto-deleted by RavenDB engine after some period of inactivity (not used by any query for N period of time)? 
index from the post above takes some time to build
querying this index may involve pagination: 
**0.0s** take first 1000 from index (first query takes time, because of index building, returns requested chunk)
**10.0s** skip 1000 take 1000 from index (lightning fast, as index is already prepared)
**11.0s** take some time exploring results from UI
**37.0s** skip 2000 take 1000 from index (still, index is expected to be prepared)
**50.0s** done with results
**some time later** _index is deleted due to the period of inactivity_


remarks:
- it is sad, that it is impossible to express complex query via RQL string only, thus making it impossible to skip explicit index creation step and rely on auto map-reduce index from query
- it is not handy that `AbstractIndexCreationTask<>` does not support parameters, making `IndexDefinition` + `string.Format()` the last resort for creating a parametrized index

**Comment by alex-bermishev** [comment]
regarding index TTL, I've found 
- [Indexing.TimeToWaitBeforeMarkingAutoIndexAsIdleInMin](https://ravendb.net/docs/article-page/5.1/http/server/configuration/indexing-configuration#indexing.timetowaitbeforemarkingautoindexasidleinmin)
- [Indexing.TimeToWaitBeforeDeletingAutoIndexMarkedAsIdleInHrs](https://ravendb.net/docs/article-page/5.1/http/server/configuration/indexing-configuration#indexing.timetowaitbeforedeletingautoindexmarkedasidleinhrs)

but this applies to auto-indexes only
I guess, I need same level of control for manually created indexes

**Reply by garayx** [reply]
static indexes are not supposed to disappear from the server, you can explicitly delete the index if you wish,  https://ravendb.net/docs/article-page/5.1/csharp/client-api/operations/maintenance/indexes/delete-index

**Reply by alex-bermishev** [reply]
static index query results can be fetched by the client with pagination over the period of time
so, there is no way to instruct RavenDB server to do garbage collection of specific static index after "it has not been used for N period of time"?
I will need to keep track of such static indexes and delete them manually?

it would be great, if there is a way to create index manually with `Type == IndexType.AutoMapReduce` and let RavenDB server do the job =)

**Reply by ayende** [reply]
Creating an index automatically yourself is not really something that you should do.
You are better off indexing the relevant fields and then using `facets` for the computation.

**Reply by alex-bermishev** [reply]
@ayende, I've tried to examine `facets` feature in order to get the results like in my monstrous map/reduce manual index =)
`facets` are about aggregation functions, min/max/avg..., and I don't get the idea how it could help 
like in this sample [https://demo.ravendb.net/demos/faceted-search/facets-basics](https://demo.ravendb.net/demos/faceted-search/facets-basics)
I need actual document records in the result, filtered by dynamic predicate, grouped, taken first and last in each group, not an aggregated value

**Comment by georgiosd** [comment]
@alex-bermishev I'm not sure I understand why you want to create an index for this - you can just have one index that indexes the timestamps and identifiers and then work with facets to get the min/max like Oren suggested. 

You would make the calculation of the entry/exit in memory or do two lazy queries with order by asc/desc to get min/max.

Wouldn't that be a good solution?

**Reply by alex-bermishev** [reply]
ok, in the real world we have hundreds of thousands `subjects` moving around thousands of `territories`
which makes millions of records over the period of time
business task is to get fist entry and last exits for specified `subjects` at given `territories` for specified period of time (like last 6 months)

- in-memory computation with some generic index will lead to fetching lots and lots of records from the RavenDB, grouping, finding min/max record by timestamp in-memory to get a result of 10k records from 5mln fetched (for example). huge memory spike, heavy serialization/deserialization load. I am certain, that it is not an accepted solution

- yes, it is possible to get the result via several queries, involving `document id` to get actual records at the end, but getting same result via single operation is much more efficient and atomic

my goal is to try to utilize functionality of RavenDB as much as possible, doing all possible filtering, calculation and projection within engine

**Reply by alex-bermishev** [reply]
one more thing: user will consume query result in paged manner. any in-memory computation will lead to keeping track of queries, sessions, results caching, ttl, etc. really messy and unreliable solution
having a possibility to simply use `limit` & `offset` will make life so much easier =)

**Reply by ayende** [reply]
Creating indexes and deleting them on the fly is _not_ recommended.
The use of `facets()` is how we recommend to resolve this.

**Reply by alex-bermishev** [reply]
> Creating indexes and deleting them on the fly is _not_ recommended. The use of `facets()` is how we recommend to resolve this.

@ayende, I am playing with `facet()` for some time and still dont get the idea how it can be used to get the desired result

[https://ayende.com/blog/169985/excerpts-from-the-ravendb-performance-team-report-facets-of-information-part-i](https://ayende.com/blog/169985/excerpts-from-the-ravendb-performance-team-report-facets-of-information-part-i)
> Facets are annoying.

=)

all `facet()` usage examples show getting aggregated values over some predicate
I need to:
- filter document collection by predicate (multiple terms)
- group filtered by `subject_id`
- get first (or last) document from each group ordered by `timestamp`

I am playing with this:
index
```
from e in docs.events
from identifier in e.Payload.identifiers
select new 
{
    HeaderID = e.Header.ID,
    HeaderTimestamp = e.Header.Timestamp,
    HeaderName = e.Header.Name,
    TerritoryID = e.Payload.territory_id,
    SubjectID = identifier.subject_id
}
```

query
```
from index "events/facet_test" 
where
    ( HeaderTimestamp between "2000-01-01T00:00:00.000+03:00" and "2000-01-05T00:00:00.000+03:00" )
    and ( SubjectID in ( "0847fada-b12e-45d0-aca5-43917457b0bc", "0054ab80-50aa-4012-808e-babcd10fdf05" ) )
    and ( HeaderName == "territory_entry" )
select 
    facet( SubjectID, max( HeaderTimestamp ) )
```
results are like this:
```JSON
{
    "Name": "SubjectID",
    "Values": [
        {
            "Name": "HeaderTimestamp",
            "Max": 0,
            "Count": 5,
            "Range": "0054ab80-50aa-4012-808e-babcd10fdf05"
        },
        {
            "Name": "HeaderTimestamp",
            "Max": 0,
            "Count": 2,
            "Range": "0847fada-b12e-45d0-aca5-43917457b0bc"
        }
    ],
    "RemainingHits": 0,
    "RemainingTermsCount": 0,
    "RemainingTerms": [],
    "@metadata": {
        "@attachments": [],
        "@counters": [],
        "@timeseries": []
    }
}
```
no actual document id is incorporated in the result (which could allow to retrieve actual documents)
_notice "Max": 0, which is a clearly strange value given `HeaderTimestamp` is `DateTimeOffset`_

I am really stuck with this `facet()` concept right now =(

could you please give me a hint to go in the right direction?
or lead to some examples with complex usage of facets for groping and getting actual documents?

sorry for being annoying, I just want to get things done right and understand how it actually works =)
thank you

**Reply by georgiosd** [reply]
`yes, it is possible to get the result via several queries, involving document id to get actual records at the end, but getting same result via single operation is much more efficient and atomic`

Have you measured it? The two queries would run on the same result set in parallel. We generate a dashboard with about ~10 lazy queries in just a few ms.

Other than that, it would help to have more context on your actual problem. How often are these queries generated? Are the time intervals predetermined? etc.

**Reply by ayende** [reply]
Facets operate only on numeric values. You need to add:

`    HeaderTimestampTicks = e.Header.Timestamp.Ticks,`

And then: `    facet( SubjectID, max( HeaderTimestampTicks ) )`

**Reply by alex-bermishev** [reply]
given millions of documents like this:

```JSON
{
    "Header": {
        "ID": "1720fba6-4a99-4213-9414-9c23b886aa9a",
        "Timestamp": "2000-03-10T10:39:00.000+03:00",
        "User": "$root",
        "Sender": "test/territory/event_generator",
        "Name": "territory_exit",
        "Schema": "territory"
    },
    "Payload": {
        "territory_id": "02d57053-09b3-42d0-b205-b5d9b384bb88",
        "identifiers": [
            {
                "code": "",
                "subject_id": "f8f48061-5d5a-4070-847c-a839b6b425eb"
            },
            {
                "code": "",
                "subject_id": "1b90a88d-c486-4fec-a7fa-0ac03a5a6e3e"
            }
        ]
    }
}
````

we need to get first entry and last exit document of each subject with filters, provided by user

- filter by `Header.Timestamp` interval (required, provided by user from UI)
- filter by `Header.Name`  IN ("territory_entry", "territory_exit") (required, business constraint)
- filter by `Payload.territory_id` IN ( ... ) (optional, list of territories provided by user from UI)
- filter by `Payload.identifiers[].subject_id` IN ( ... ) (optional, list of subjects provided by user from UI)
- group by `Payload.identifiers[].subject_id`
- get **first** document in each group (by `Header.Timestamp`), where `Header.Name` == "territory_entry"
- get **last** document in each group (by `Header.Timestamp`), where `Header.Name` == "territory_exit"

ideal result would be array of entry/exit document projections by each subject_id
```JSON
[
  {
      "SubjectID": "57d68a46-8af1-4287-882e-e46ea75f6418",
      "Entry": {
          "ID": "a111aae2-b099-4fae-8b14-48ee675d60b7",
          "Name": "territory_entry",
          "Timestamp": "2000-01-01T04:08:00.0000000",
          "TerritoryID": "61fafd9d-6aa3-4da8-975d-e707a5aa0612",
          "Identifier": {
              "code": "",
              "subject_id": "57d68a46-8af1-4287-882e-e46ea75f6418"
          }
      },
      "Exit": {
          "ID": "f084f062-e435-4382-8ea0-d358cc5096d4",
          "Name": "territory_exit",
          "Timestamp": "2000-01-01T13:32:00.0000000",
          "TerritoryID": "cae6a730-6e9a-4046-af9d-0b17d258309d",
          "Identifier": {
              "code": "",
              "subject_id": "57d68a46-8af1-4287-882e-e46ea75f6418"
          }
      }
  }
]
```

all predicate parameters (except `Header.Name`  IN ("territory_entry", "territory_exit") ) are taken from user input
query execution is not really frequent, 10+ queries per day average
query result must be able to be retrieved with paging (user will examine the result page by page)

I managed to get this result with manual index creation, but if it will be more efficient to use other approaches as already mentioned `facet()` (even if I have no idea how to use it for this task right now =) ), any help will be greatly appreciated

**Reply by alex-bermishev** [reply]
thanks @ayende for `.Ticks` hint. that works

@georgiosd, I can get same result with generic predefined index like so
```
from index "events/test" 
where
    ( HeaderTimestampTicks between $timestamp_from and $timestamp_to )
    and ( SubjectID == $subject_id )
    and ( HeaderName == "territory_entry" )
order by HeaderTimestampTicks
limit 1

from index "events/test" 
where
    ( HeaderTimestampTicks between $timestamp_from and $timestamp_to )
    and ( SubjectID == $subject_id )
    and ( HeaderName == "territory_exit" )
order by HeaderTimestampTicks desc 
limit 1
```

running 2 queries for **each** (sic!) `subject_id` from the user input filter
but if user selects 1k (for example) subjects to filter, that will require 1k * 2 queries to get the result, cache the result for paging, keep track of this particular request TTL to expire and garbage collect result (user examined several pages and quit, or issue another query request) 
really bad solution for me

**Reply by georgiosd** [reply]
Yes facets is much more tidy. 

Alas, is the user actually likely to select 1k subjects to filter? Are there no other assumptions you can make on the user input?

**Reply by alex-bermishev** [reply]
`subjects` belong to `departments`, as of buisness layer
user can select `departments` and/or individual `subjects` to filter
`department` can contain thousands of `subjects`
user filter gets unfolded to the collection of `subject_ids` for now, as actual `events` in RavenDB have no knowledge of `departments` and do not operate this term
of course, it is a subject of optimization which I am looking forward to later
anyway, even after this optimization, we can have lots of separate queries to make (if not thousands, but dozens - for sure)
also, I do not want to manage results caching, expiration, etc. for pagination purposes
